modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/sqlquery/DbExecutor.scala

```diff
@@ -81,26 +83,8 @@ object DbExecutor {
 
   implicit def syncDbExecutor[F[_]: Sync]: DbExecutor[F] =
     new DbExecutor[F] {
-      def getConnection(rdbms: Rdbms, connectionRef: ConnectionRef[F])(implicit M: Monad[F]): F[Either[Throwable, Connection]] =
-        for {
-          cachedConnection <- connectionRef.get(()).map(flattenCached)
-          connection <- cachedConnection match {
-                          case Right(conn) =>
-                            for {
-                              closed <- Sync[F].delay(conn.isClosed)
-                              result <- if (!closed) conn.asRight[Throwable].pure[F]
-                                        else
-                                          for {
-                                            newConn <- Sync[F].delay(
-                                                         Either.catchNonFatal(DriverManager.getConnection(rdbms.connectionString))
-                                                       )
-                                            _ <- connectionRef.put((), newConn)
-                                          } yield newConn
-                            } yield result
-                          case Left(error) =>
-                            error.asLeft[Connection].pure[F]
-                        }
-        } yield connection
+      def getConnection(dataSource: HikariDataSource): F[Either[Throwable, Connection]] =
```

Comment:
I can see you are being very careful to catch every possible exception, and close the connection if anything goes wrong.  However, it might be easier to change the `DbExecutor` trait so that instead of `getConnection` it has `withConnection`.  This can ensure that the exception is always closed, and then you don't need to be so careful about it in other places in the code.

```
implicit def syncDbExecutor[F[_]: Sync]: DbExecutor[F] =
  new DbExecutor[F] {
    def withConnection[A](dataSource: HikariDataSource)(f: Connection => F[Either[Throwable, A]]): F[Either[Throwable, A]] =
      Sync[F].bracket(Sync[F].delay(dataSource.getConnection()))(f)(cx => Sync[F].delay(cx.close())
        .handleError(e => Left(e))
```

```
implicit def idDbExecutor: DbExecutor[Id] =
  new DbExecutor[Id] {
    def  withConnection[A](dataSource: HikariDataSource)(f: Connection => Either[Throwable, A]): Either[Throwable, A] =
      try {
        // get connection and use connection
      } finally {
        // close connection
     }
```

Just an idea.

Reply:
Please see https://github.com/snowplow/enrich/pull/481/commits/64cc9585c2c7f3d145776af31be12e43fb71f821